### PR TITLE
Uart ble example

### PR DESCRIPTION
**Pull request by wolfc01**
Hello 

I have created a full example of flutter_reactive_ble, for the use-case of emulating an "nordic ble UART" example, which is used by a lot of experimenters creating BLE solutions together with a pyboard, pycom and arduino boards.

I hope this example will help this ble library to grow. 

Please let me know if alterations will be necessary for accepting this pull request.

Best regards
Carl.

### CODE REVIEW

**Comment by werediver** [comment]
It's okay to have multiple _meaningful_ commits with _meaningful_ commit messages in a pull-request, but in this case I am suggesting to squash all of them, if this pull request is to be merged.

File: uart_example/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# flutter reactive BLE UART demo app
```
Comment (werediver) [inline uart_example/README.md:1]: Sentence capitalization, please: "# Flutter ..."

File: uart_example/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+This demo app shows an usecase for using flutter_reactive_ble library for building a UART emulation over BLE.
```
Comment (werediver) [inline uart_example/README.md:3]: Please, spell "use case" as two separate words and correct the article: "a use case", not ~"an use case"~ ("u" here reads as /ˈjuː/, the first _sound_ is consonant).

File: uart_example/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+This demo app uses one service: the Nordic UART Service with vendor-specific UUID:
+
+```
```
Comment (werediver) [inline uart_example/README.md:7]: Please, specify the language in all code-blocks (\```dart)

File: uart_example/README.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+Uuid _UART_RX   = Uuid.parse("6E400002-B5A3-F393-E0A9-E50E24DCCA9E");
+Uuid _UART_TX   = Uuid.parse("6E400003-B5A3-F393-E0A9-E50E24DCCA9E");
+```
+
+## Usage
+
+This app can be used if the BLE peripheral implements the aforementioned services. The user interface is shown below:
+
+![alt text](./screenshot.png "User interface")
```
Comment (werediver) [inline uart_example/README.md:22]: "alt text" is not a useful alt text and "screenshot.png" is only useful if it's the only screenshot here (which it probably is, so fine).

Consider replacing "alt text" with something like "Application screenshot".

File: uart_example/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+Use the play button for scanning and the stop button for stop a running scan. When an UART BLE service is found, it is displayed at the top. Press the found device at the right, then a connection is being established as can be seen in the Status messages pane. Then a message string can be sent to the BLE peripheral, Received data is shown in the Received data pane. An existing connection can be closed using the bottom rightmost button.
+
+## BLE undeliverable exception
```
Comment (werediver) [inline uart_example/README.md:26]: I am not sure it is useful to duplicate this section here. I'd rather suggest to either refer to the top-level README or omit entirely.

File: uart_example/android/app/build.gradle [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    sourceSets {
+        main.java.srcDirs += 'src/main/kotlin'
+    }
+
+    lintOptions {
+        disable 'InvalidPackage'
+    }
+
+    defaultConfig {
+        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
```
Comment (werediver) [inline uart_example/android/app/build.gradle:40]: "TODO" means "technical debt". If this is a hint to whoever is building this example app, I wouldn't call it a TODO.

Same below (currently at line 50).

File: uart_example/lib/main.dart [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+  void _stopScan() async {
+    await _scanStream.cancel();
+    _scanning = false;
+    refreshScreen();
+  }
+
+  Future<void> showNoPermissionDialog() async {
+    return showDialog<void>(
+      context: context,
+      barrierDismissible: false, // user must tap button!
```
Comment (werediver) [inline uart_example/lib/main.dart:91]: I don't think this comment is useful. Yes, it explains the meaning of the parameter in words simpler than the parameter name (`barrierDismissible:`), but are you going to put such comments on every line now?

File: uart_example/lib/main.dart [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+      refreshScreen();
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        // Here we take the value from the MyHomePage object that was created by
+        // the App.build method, and use it to set our appbar title.
```
Comment (werediver) [inline uart_example/lib/main.dart:196]: The next line explains it in four "words" of code, I'd say. Consider removing the comment.

File: uart_example/lib/main.dart [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+                )
+              ),
+              height: 100,
+              child: ListView.builder(
+                  itemCount: _foundBleUARTDevices.length,
+                  itemBuilder: (context, index) {
+                    return Card(
+                        child: ListTile(
+                          dense: true,
+                          enabled: !((!_connected && _scanning) || (!_scanning && _connected)),
```
Comment (werediver) [inline uart_example/lib/main.dart:220]: `!a && b || a && !b` is equivalent to `a != b` (exclusive OR) and `!(a != b)` is `a == b`, so this whole expression should be equivalent to `_scanning == _connected`, if I didn't miss anything.

File: uart_example/lib/main.dart [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+                            child: Container(
+                              width: 48,
+                              height: 48,
+                              padding: EdgeInsets.symmetric(vertical: 4.0),
+                              alignment: Alignment.center,
+                              child: Icon(Icons.add_link),
+                            ),
+                          ),
+                          subtitle: Text(_foundBleUARTDevices[index].id),
+                          title: Text(index.toString() + ":" +" "+_foundBleUARTDevices[index].name),
```
Comment (werediver) [inline uart_example/lib/main.dart:235]: Use string interpolation instead.

File: uart_example/pubspec.yaml [hunk 11]
```diff
@@ -0,0 +1,78 @@
+name: ble_testapp
+description: flutter ble testapp
+
+# The following line prevents the package from being accidentally published to
+# pub.dev using `pub publish`. This is preferred for private packages.
```
Comment (werediver) [inline uart_example/pubspec.yaml:5]: Please, clean-up the comments in this file. I understand they are generated automatically, but they are not useful (they can't replace the documentation, but cluttering the view).

File: uart_example/test/widget_test.dart [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+// gestures. You can also use WidgetTester to find child widgets in the widget
+// tree, read text, and verify that the values of widget properties are correct.
+
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import 'package:ble_testapp/main.dart';
+
+void main() {
+  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
```
Comment (werediver) [inline uart_example/test/widget_test.dart:14]: This had nothing to do with the actual implementation, right? An unrelated test should be removed (the whole file).

**Review by werediver — COMMENTED** [review]
I've just glanced over the pull-request, there probably are more things to adjust.

**Comment by werediver** [comment]
My main concern regarding this pull-request is that it introduces a whole new example application. The official documentation (implicitly) suggests there should be a single example app in a package ([ref.](https://flutter.dev/docs/development/packages-and-plugins/developing-packages#step-1-create-the-package-1)).

The use case covered in this demo app may be useful, but I suggest considering integrating it into the main example app (by extending, reworking, or even replacing, whatever works best).

CC @remonh87

**Comment by farr64** [comment]
@remonh87 wrote:
> My main concern regarding this pull-request is that it introduces a whole new example application. The official documentation (implicitly) suggests there should be a single example app in a package ([ref.](https://flutter.dev/docs/development/packages-and-plugins/developing-packages#step-1-create-the-package-1)).
> 
> The use case covered in this demo app may be useful, but I suggest considering integrating it into the main example app (by extending, reworking, or even replacing, whatever works best).

I agree: The current example app is very good and I believe, given my limited knowledge and skill at this point, that it can be (relatively easily) extended with the three items that I mentioned in #170.

First of all, Carl, thanks very much for the enthusiasm. 

I'm a total newcomer to Flutter but I'm very interested in exploring and learning. So, my heartfelt thanks to everyone who is contributing to this worthwhile effort. I hope I'll be able to contribute as soon as I come up to speed. Right now, I'm still at the starting blocks (blocked by lack of knowledge and skill, which are easily removed ;-)

**Comment by remonh87** [comment]
@wolfc01  many thanks for creating this pull request we are happy to see new functionalities added to the example app. I have to agree with @werediver  I think we should integrate the functionality in the current example app. If you feel like you need some help with this feel free to reach out.

**Comment by wolfc01** [comment]
> My main concern regarding this pull-request is that it introduces a whole new example application. The official documentation (implicitly) suggests there should be a single example app in a package ([ref.](https://flutter.dev/docs/development/packages-and-plugins/developing-packages#step-1-create-the-package-1)).
> 
> The use case covered in this demo app may be useful, but I suggest considering integrating it into the main example app (by extending, reworking, or even replacing, whatever works best).
> 
> CC @remonh87

Hi @werediver,

thanks for your comments. I understand that it is preferred to have one example. However, my example is quite specific to "Nordic UART over ble" example. If you have no such peripheral device available, the example will not be quite useful.

How about this @werediver and @remonh87 ?

Br
Carl

**Comment by remonh87** [comment]
@wolfc01  After more carefully checking your example is that it is indeed a very specific case. What I will propose it to add  a link in  our readme to your repository.  I think that would fit very nicely and later we can add more example usages.. What do you think?

**Comment by werediver** [comment]
I support Remon's suggestion. This demo app needs it's own repo and we can mention it in README.

**Comment by wolfc01** [comment]
> What I will propose it to add a link in our readme to your repository. I think that would fit very nicely and later we can add more example usages.. What do you think?

That's is what I would prefer. 

> After more carefully checking your example is that it is indeed a very specific case.
Indeed. IMHO this applies for all examples regarding BLE. Therefore my suggestion to end up with a vast array  of examples in a joint massive effort ;-)

Br
Carl

**Comment by wolfc01** [comment]
> I support Remon's suggestion. This demo app needs it's own repo and we can mention it in README.

Good, nice. If you can create a repo, I will issue an pull request against it after processing @werediver review comments?

Br
Carl

**Comment by werediver** [comment]
@wolfc01 I see no reason why your work should go to [PhilipsHue/](https://github.com/PhilipsHue/) space, if that's what you mean.

Please, go ahead and create your own repo. If you wish, you may grant some extra rights to Remon and me, but that is not required.

File: uart_example/README.md (hunk 5 above)
Comment (wolfc01) [reply uart_example/README.md:26]: Well, it took me quite some time to fix the BLE undeliverable exception, the top-level README is a bit limited on this topic for somebody who does not know android programming at all. So an example would help here.

**Comment by wolfc01** [comment]
Good, i will create my own repo and would be happy if:
* the code is reviewed for correct usage of the flutter_reactive_ble library
* a reference is created to it in your main readme.md

Agree?

Br
Carl.

**Comment by wolfc01** [comment]
As agreed this example will live in an own repository, so I close the pull request.

**Pull request closed without merging**